### PR TITLE
Expression matrix zarr processing

### PR DESCRIPTION
**Pull request by keller-mark**
This adds a script to process the `cluster_marker_genes.h5ad` files into the Zarr expression matrix format, and also updates the annotation CSV file in the source-data directory (sent by the Satija Lab).

### CODE REVIEW

File: snakemake/satija/Snakefile [hunk 1]
```diff
@@ -52,11 +53,25 @@ rule process_dataset:
...
         '''
 
+rule convert_h5ad_to_zarr:
+    input:
+        (RAW_DIR / "uf-processed" / "{globus_id}" / "cluster_marker_genes.h5ad")
+    output:
+        directory(PROCESSED_DIR / "{globus_id}.expression-matrix.zarr")
+    params:
+        script=(SRC_DIR / "convert_h5ad_to_zarr.py")
+    shell:
```
Comment (mccalluc) [inline snakemake/satija/Snakefile:63]: Would it be weird to use `run` and execute the python code directly, without the shell getting involved?

File: snakemake/satija/src/convert_h5ad_to_zarr.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    gexp = read_h5ad(input_file)
+    gexp_arr = gexp.X
+    gexp_df = gexp.to_df()
+
+    # Re-scale the gene expression values between 0 and 255 (one byte ints).
+    gexp_arr_min = gexp_arr.min()
+    gexp_arr_max = gexp_arr.max()
+    gexp_arr_range = gexp_arr_max - gexp_arr_min
+    gexp_arr_ratio = 255 / gexp_arr_range
+    gexp_norm_arr = (gexp_arr - gexp_arr_min) * gexp_arr_ratio
```
Comment (mccalluc) [inline snakemake/satija/src/convert_h5ad_to_zarr.py:20]: Is it still floats at this point?

**Review by mccalluc — APPROVED** [review]
Has the structure of the data from Rahul changed? If it hasn't, I wouldn't change the fixtures unnecessarily, but not something to worry about now.

Wasn't clear to me when / if the conversion to ints happens, but not sure it matters, either.

Thanks!

File: snakemake/satija/src/convert_h5ad_to_zarr.py (hunk 2 above)
Comment (ilan-gold) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: I noticed this too when looking back at the other PR - I saw some sliders on the Vitessce PR so how does that work?

**Review by ilan-gold — APPROVED** [review]

File: snakemake/satija/src/convert_h5ad_to_zarr.py (hunk 2 above)
Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: Yes the casting does not happen until the `zarr.open` line where I pass `dtype='uint8'`

Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: I did add a color scale slider, it works fine with the integer data

Comment (ilan-gold) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: I am wondering then why not take it further and use uint16 or a 32 bit float (i.e scaling to 0-1) if we have a slider.  Is this standard in other heat maps?

File: snakemake/satija/Snakefile (hunk 1 above)
Comment (keller-mark) [reply snakemake/satija/Snakefile:63]: The shell option is nice in that it forces you to write standalone scripts with argparse that are not coupled to snakemake (of course at the cost of writing the argparse code).

The `script` and `run` options couple code to snakemake slightly more, although the main logic can / should always be put in regular python functions anyway.

File: snakemake/satija/src/convert_h5ad_to_zarr.py (hunk 2 above)
Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: Wouldn't that double or quadruple the file sizes though? These zarr directories are ~15 MB and I am requesting all chunks at the same time in the browser

Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: I just tried increasing to uint16 for one of the datasets and file size increased from 15.4 MB to 22.9 MB

Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: As far as other heatmaps:
- clustergrammer: TSV or in-memory pandas df via python https://clustergrammer.readthedocs.io/matrix_format_io.html
    - the largest file I see in their example files is a 3.1 MB TSV with shape (3406, 458) https://github.com/MaayanLab/clustergrammer/blob/master/txt/KEA_matrix.txt

- plotly
    - there is an example of a heatmap that takes a JPG image as input https://plotly.com/javascript/heatmap-webgl/#webgl-heatmaps
    - there are other examples here https://plotly.com/javascript/heatmaps/

- sanddance: loads a 3.59 MB TSV https://sanddance.js.org/app/

I don't see any interactive color scales

**Comment by keller-mark** [comment]
The data structure from Rahul has not changed, just the file name (it has a new date). I removed the date suffix from the file and directory names, now it only exists in the source data download URL

File: snakemake/satija/src/convert_h5ad_to_zarr.py (hunk 2 above)
Comment (keller-mark) [reply snakemake/satija/src/convert_h5ad_to_zarr.py:20]: This is a demo with the color scale slider https://s3.amazonaws.com/vitessce-data/demos/2020-07-21/0ccb14d/index.html?dataset=linnarsson-2018&theme=dark

**Pull request merged**